### PR TITLE
Toni day 1

### PR DESCRIPTION
**Pull request by tonilopezmr**
# Toni's solution for day 1 challenge

## Description

In the first part, I only add in the last of the list a first number and with a fold I compare each pair number.

And in the second part , I add the length / 2 first numbers at the end of the sequence and compare `actual index + (length / 2)`

## Solution

Part 1

```kotlin
fun sequence(sequence: String) : Int {
        val captcha = (sequence + sequence[0])
        
        return captcha.foldIndexed(0) { idx, acc, a ->
            if (idx < sequence.length && a.toString().toInt() == captcha[idx +1].toString().toInt()) {
                acc + a.toString().toInt()
            } else {
                acc
            }
        }
    }
```

Part2 (same as the part 1 but with displacement)

```kotlin
fun sequence(sequence: String) : Int {
        if (sequence.length % 2 != 0) return 0

        val displacement = sequence.length / 2
        val captcha = sequence + sequence.substring(0, displacement)

        return captcha.foldIndexed(0) { idx, acc, a ->
            if (idx < sequence.length && a.toString().toInt() == captcha[idx + displacement].toString().toInt()) {
                acc + a.toString().toInt()
            } else {
                acc
            }
        }
    }
```

## Tests battery

In this case the test are trivial, but for example I have tested the input length of the second part which return 0 if the sequence is not even. 

```
 @Test
    fun `the sequence number must be even`() {
        assertEquals(0, sequence("1"))
    }
```

A test example:

```kotlin
@Test
    fun `1212 produces 6`() {
        assertEquals(6, sequence("1212"))
    }
```

Happy Advent of Code!

### CODE REVIEW

File: tonilopezmr/CaptchaPartOne.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.junit.Test
+import org.junit.Assert.*
+
+
+
+class CaptchaPartOne {
+
+    fun sequence(sequence: String) : Int {
+        val captcha = (sequence + sequence[0])
```
Comment (JoseLlorensRipolles) [inline tonilopezmr/CaptchaPartOne.kt:11]: Uhh I like this better than:

if sequence[0] == sequence[-1]:
    acc += sequance[0]

:+1:

**Review by JoseLlorensRipolles — APPROVED** [review]

File: tonilopezmr/CaptchaPartOne.kt (hunk 1 above)
Comment (tonilopezmr) [reply tonilopezmr/CaptchaPartOne.kt:11]: What? do you prefer a if? that simple sum?

File: tonilopezmr/CaptchaPartTwo.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+class CaptchaPartTwo {
+
+    fun sequence(sequence: String) : Int {
+        if (sequence.length % 2 != 0) return 0
+
+        val displacement = sequence.length / 2
+        val captcha = sequence + sequence.substring(0, displacement)
+
+        return captcha.foldIndexed(0) { idx, acc, a ->
+            if (idx < sequence.length && a.toString().toInt() == captcha[idx + displacement].toString().toInt()) {
```
Comment (JoseLlorensRipolles) [inline tonilopezmr/CaptchaPartTwo.kt:16]: What about captcha[(idx  + displacement)%size] ? That way you don't need line 13, although you do 1 more operation each iteraation 🤔

File: tonilopezmr/CaptchaPartOne.kt (hunk 1 above)
Comment (JoseLlorensRipolles) [reply tonilopezmr/CaptchaPartOne.kt:11]: Nono I say I like it **better** 😄

File: tonilopezmr/CaptchaPartOne.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        return captcha.foldIndexed(0) { idx, acc, a ->
+            if (idx < sequence.length && a.toString().toInt() == captcha[idx +1].toString().toInt()) {
+                acc + a.toString().toInt()
+            } else {
+                acc
+            }
+        }
+    }
```
Comment (Cotel) [inline tonilopezmr/CaptchaPartOne.kt:19]: At first I though of a Fold/Zip solution too but it involved working with indexes and I'm not a big fan of it.

Just one thing about this method. You could map all strings to int before folding, that way you would avoid calling `x.toString().toInt()` 3 times. Casting is one of the most expensive operations, but in this case it wouldn't impact that much on performance probably.

File: tonilopezmr/CaptchaPartTwo.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        assertEquals(12, sequence("123123"))
+    }
+
+    @Test
+    fun `12131415 produces 4`() {
+        assertEquals(4, sequence("12131415"))
+    }
+
+    @Test
+    fun `the result of the day 1 part 2 is`() {
```
Comment (Cotel) [inline tonilopezmr/CaptchaPartTwo.kt:55]: Nice test 👍

File: tonilopezmr/CaptchaPartTwo.kt (hunk 2 above)
Comment (tonilopezmr) [reply tonilopezmr/CaptchaPartTwo.kt:16]: Yes! That's right!

File: tonilopezmr/CaptchaPartOne.kt (hunk 3 above)
Comment (tonilopezmr) [reply tonilopezmr/CaptchaPartOne.kt:19]: Yes! I know but I don't know in a big input or big operation what is better, if two iterations or casting 🤔 

Maybe casting

File: tonilopezmr/CaptchaPartTwo.kt (hunk 4 above)
Comment (tonilopezmr) [reply tonilopezmr/CaptchaPartTwo.kt:55]: First time I run this test I expect 0 in the result when test fails with the real result I copy to expect number hahaha

File: tonilopezmr/CaptchaPartOne.kt (hunk 3 above)
Comment (Cotel) [reply tonilopezmr/CaptchaPartOne.kt:19]: In this case the difference between casting or iterating one more time won't make a difference 🤷‍♂️

**Pull request merged**